netann/chan_status_manager.go

```diff
@@ -0,0 +1,595 @@
+//   ChanEnableTimeout + ChanStatusSampleInterval > ChanDisableTimeout.
+func NewChanStatusManager(cfg *ChanStatusConfig) (*ChanStatusManager, error) {
+	// Assert that the config timeouts are properly formed. We require the
+	// enable_timeout + sample_interval to be less than or equal to the
+	// disable_timeout and that all are positive values. A peer that
+	// disconnects and reconnects quickly may cause a disable update to be
+	// sent, shortly followed by a reenable. Ensuring a healthy separation
+	// helps dampen the possibility of spamming updates that toggle the
+	// disable bit for such events.
+	if cfg.ChanStatusSampleInterval <= 0 {
+		return nil, ErrInvalidTimeoutConstraints
+	}
+	if cfg.ChanEnableTimeout <= 0 {
+		return nil, ErrInvalidTimeoutConstraints
+	}
+	if cfg.ChanDisableTimeout <= 0 {
+		return nil, ErrInvalidTimeoutConstraints
+	}
+	if cfg.ChanEnableTimeout+cfg.ChanStatusSampleInterval >
+		cfg.ChanDisableTimeout {
+		return nil, ErrInvalidTimeoutConstraints
+
+	}
+
+	return &ChanStatusManager{
+		cfg:                cfg,
+		ourPubKeyBytes:     cfg.OurPubKey.SerializeCompressed(),
+		chanStates:         make(channelStates),
+		statusSampleTicker: time.NewTicker(cfg.ChanStatusSampleInterval),
+		enableRequests:     make(chan statusRequest),
+		disableRequests:    make(chan statusRequest),
+		quit:               make(chan struct{}),
+	}, nil
+}
+
+// Start safely starts the ChanStatusManager.
+func (m *ChanStatusManager) Start() error {
+	if !atomic.CompareAndSwapUint32(&m.started, 0, 1) {
+		return nil
+	}
+
+	channels, err := m.fetchChannels()
+	if err != nil {
+		return err
+	}
+
+	// Populate the initial states of all confirmed, public channels.
+	for _, c := range channels {
+		_, err := m.getOrInitChanStatus(c.FundingOutpoint)
+		switch {
+
+		// If we can't retrieve the edge info for this channel, it may
+		// have been pruned from the channel graph but not yet from our
+		// set of channels. We'll skip it as we can't determine its
+		// initial state.
+		case err == channeldb.ErrEdgeNotFound:
+			log.Warnf("Unable to find channel policies for %v, "+
+				"skipping. This is typical if the channel is "+
+				"in the process of closing.", c.FundingOutpoint)
+			continue
+
+		case err != nil:
+			return err
+		}
+	}
+
+	m.wg.Add(1)
+	go m.statusManager()
+
+	return nil
+}
+
+// Stop safely shuts down the ChanStatusManager.
+func (m *ChanStatusManager) Stop() error {
+	if !atomic.CompareAndSwapUint32(&m.stopped, 0, 1) {
+		return nil
+	}
+
+	close(m.quit)
+	m.wg.Wait()
+
+	return nil
+}
+
+// RequestEnable submits a request to immediately enable a channel identified by
+// the provided outpoint. If the channel is already enabled, no action will be
+// taken. If the channel is marked pending-disable the channel will be returned
+// to an active status as the scheduled disable was never sent. Otherwise if the
+// channel is found to be disabled, a new announcment will be signed with the
```

Comment:
Nit: s/announcment/announcement/. Saw some others like this in this file as well ;)

Reply:
grr, i've been wondering why my autocomplete keeps doing that, thanks